packages/react-components/react-combobox/src/components/Dropdown/useDropdown.tsx

```diff
@@ -36,72 +34,19 @@ export const useDropdown_unstable = (props: DropdownProps, ref: React.Ref<HTMLBu
 
-  // jump to matching option based on typing
-  const searchString = React.useRef('');
-  const [setKeyTimeout, clearKeyTimeout] = useTimeout();
-
-  const getNextMatchingOption = (): OptionValue | undefined => {
-    // first check for matches for the full searchString
-    let matcher = (optionText: string) => optionText.toLowerCase().indexOf(searchString.current) === 0;
-    let matches = getOptionsMatchingText(matcher);
-    let startIndex = activeOption ? getIndexOfId(activeOption.id) : 0;
-
-    // if the dropdown is already open and the searchstring is a single character,
-    // then look after the current activeOption for letters
-    // this is so slowly typing the same letter will cycle through matches
-    if (open && searchString.current.length === 1) {
-      startIndex++;
-    }
-
-    // if there are no direct matches, check if the search is all the same letter, e.g. "aaa"
-    if (!matches.length) {
-      const letters = searchString.current.split('');
-      const allSameLetter = letters.length && letters.every(letter => letter === letters[0]);
-
-      // if the search is all the same letter, cycle through options starting with that letter
-      if (allSameLetter) {
-        startIndex++;
-        matcher = (optionText: string) => optionText.toLowerCase().indexOf(letters[0]) === 0;
-        matches = getOptionsMatchingText(matcher);
-      }
-    }
-
-    // if there is an active option and multiple matches,
-    // return first matching option after the current active option, looping back to the top
-    if (matches.length > 1 && activeOption) {
-      const nextMatch = matches.find(option => getIndexOfId(option.id) >= startIndex);
-      return nextMatch ?? matches[0];
-    }
-
-    return matches[0] ?? undefined;
-  };
-
-  const onTriggerKeyDown = (ev: React.KeyboardEvent<HTMLButtonElement>) => {
-    // clear timeout, if it exists
-    clearKeyTimeout();
-
-    // if the key was a char key, update search string
-    if (getDropdownActionFromKey(ev) === 'Type') {
-      // update search string
-      searchString.current += ev.key.toLowerCase();
-      setKeyTimeout(() => {
-        searchString.current = '';
-      }, 500);
-
-      // update state
-      !open && setOpen(ev, true);
-
-      const nextOption = getNextMatchingOption();
-      setActiveOption(nextOption);
-      setFocusVisible(true);
-    }
-  };
-
-  // resolve button and listbox slot props
-  let triggerSlot: Slot<'button'>;
-  let listboxSlot: Slot<typeof Listbox> | undefined;
+  const triggerRef = React.useRef<HTMLButtonElement>(null);
+  let listbox = slot.optional(props.listbox, {
+    renderByDefault: true,
+    defaultProps: { children: props.children },
+    elementType: Listbox,
+  });
+  const listboxRef = useMergedRefs(listbox?.ref, comboboxPopupRef);
+  if (listbox) {
+    listbox.ref = listboxRef;
+  }
+  listbox = useListboxSlot(props, listbox, triggerRef) as typeof listbox;
```

Comment:
Why not following?

```ts
const listbox = useListboxSlot(props, slot.optional(props.listbox, {
    renderByDefault: true,
    defaultProps: { children: props.children },
    elementType: Listbox,
  }), triggerRef);
const listboxRef = useMergedRefs(listbox?.ref, comboboxPopupRef);

  if (listbox) {
    listbox.ref = listboxRef;
  }
```

Or even (do `slot.optional(props.listbox` in `useListboxSlot()`):

```ts
const listbox = useListboxSlot(props, triggerRef);
const listboxRef = useMergedRefs(listbox?.ref, comboboxPopupRef);

  if (listbox) {
    listbox.ref = listboxRef;
  }
```

Reply:
With some help from @bsunderhus, we figured out the interface types for slots